pallets/relay-storage-roots/src/lib.rs

```diff
@@ -0,0 +1,117 @@
+// (at your option) any later version.
+
+// Moonkit is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonkit.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Relay Storage Roots Pallet
+//!
+//! This pallet stores the latest `MaxStorageRoots` relay storage roots, which can be used to
+//! verify state proofs against an old state of the relay chain.
+//! 
+//! This is useful when the proof needs to be generated by an end user, because
+//! by the time their transaction is included in a block, the latest relay
+//! block will probably have changed and therefore the proof will be invalid.
+//! To avoid that, we expect the user to generate a proof against the latest relay block stored
+//! in this pallet. This proof will then be valid as long as that relay block is not removed from
+//! here.
+//! 
+//! This pallet SHOULD NOT be used for data that can change quickly, because we allow the user to
+//! submit a proof of an old state. Therefore a valid proof does not imply that the current relay
+//! state is the expected one.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use crate::weights::WeightInfo;
+use cumulus_pallet_parachain_system::RelaychainStateProvider;
+use cumulus_primitives_core::relay_chain::BlockNumber as RelayBlockNumber;
+use frame_support::pallet;
+use frame_support::pallet_prelude::*;
+use frame_system::pallet_prelude::*;
+pub use pallet::*;
+use sp_core::Get;
+use sp_core::H256;
+use sp_std::collections::vec_deque::VecDeque;
+
+#[cfg(any(test, feature = "runtime-benchmarks"))]
+mod benchmarks;
+pub mod weights;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+#[pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Configuration trait of this pallet.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RelaychainStateProvider: RelaychainStateProvider;
+		#[pallet::constant]
+		type MaxStorageRoots: Get<u32>;
+		/// Weight info
+		type WeightInfo: WeightInfo;
+	}
+
+	/// Map of relay block number to relay storage root
+	#[pallet::storage]
+	pub type RelayStorageRoot<T: Config> =
+		StorageMap<_, Twox64Concat, RelayBlockNumber, H256, OptionQuery>;
+
+	/// List of all the keys in `RelayStorageRoot`.
+	/// Used to remove the oldest key without having to iterate over all of them.
+	#[pallet::storage]
+	pub type RelayStorageRootKeys<T: Config> =
+		StorageValue<_, VecDeque<RelayBlockNumber>, ValueQuery>;
+
+	impl<T: Config> Pallet<T> {
+		/// Populates `RelayStorageRoot` using `RelaychainStateProvider`.
+		pub fn set_relay_storage_root() {
+			let relay_state = T::RelaychainStateProvider::current_relay_chain_state();
```

Comment:
Can you move that logic on an inherent (that will come after the parachain system inherent), this way we will be able to use the current relay block in the same block.
And generally speaking is better to not use `on_finalize` if there is another way, on_finalize` is very dangerous because it can explode the pov size limit (an inherent can too, but the client will then include less user transactions to compensate).

Reply:
My first attempt was using an inherent, but @girazoki suggested to do it `on_finalize` instead, don't remember why.

> this way we will be able to use the current relay block in the same block.

When using `on_finalize` way we will be able to prove things (because we always know the current relay chain state) but it is true that it will not be inserted in the mapping. I imagine the verification logic could handle this as an edge case, but I will not implement verification yet.

Reply:
the inherent thing seems like a bad idea to me because it forces you to modify the client.

The current storage root is likely not very useful for a user that wants to prove something, as at the time of submitting the tx her does not know which storage root the block is built upon.

Reply:
I agree with the fact the `on_finalize` is not ideal but we are using it in several places to remove storage items (i.e., all related to inherents).

I envision inherents for things that need to be proven yes or yes by the collator. This is not such a case AFAIK...

Reply:
ok that make sense, thank you for the explanation